### PR TITLE
Initial rocMLIR support in Tuna

### PR DESCRIPTION
**Pull request by pcf000**
First batch of code for rocMLIR support in Tuna.  It encompasses most if not all of rocMLIR ticket #984, which itself incorporates eleven smaller tickets;  it's the second step in rocMLIR ticket [#910](https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/910).

It's midway in complexity between example and miopen.  I can run convolution tuning in a fairly controlled environment on a lockhart node and it all works.  I've run the recommended pylint (including rocmlir/*.py).

There are two commits;  one is mostly typo correction and improvements to the documentation that I stumbled across.

### CODE REVIEW

File: doc/src/TuningCycle.md [hunk 1]
```diff
@@ -34,14 +34,14 @@ the tuning part.
 Tuning Steps
 ------------
 
-Tuning stores intermittent data in a central mySQL database. Each reference to a table, 
+Tuning stores intermittent data in a central mySQL database. Each reference to a table,
```
Comment (pcf000) [inline doc/src/TuningCycle.md:37]: My editor trims trailing whitespace.

File: tuna/miopen/subcmd/export_db.py [hunk 2]
```diff
@@ -49,8 +49,11 @@
 
 DIR_NAME = {'F': 'Fwd', 'B': 'BwdData', 'W': 'BwdWeights'}
 
-_, ID_SOLVER_MAP = get_id_solvers()
+ID_SOLVER_MAP = None
```
Comment (pcf000) [inline tuna/miopen/subcmd/export_db.py:52]: Changes to this file are because it is included by tuna/lib_utils.py and the act of importing will call get_id_solvers() because it's at top level.  get_id_solvers() looks in the tuna.solvers table and barfs if there isn't one.  In the rocmlir case, there isn't one.  These changes arrange to not initialise ID_SOLVER_MAP until it is used, and I'm sure there's a more elegant way to do it.

File: tuna/rocmlir/import_configs.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+###############################################################################
+""" Module for tagging and importing configs """
```
Comment (pcf000) [inline tuna/rocmlir/import_configs.py:27]: Adapted from tuna/miopen/subcmd/import_configs.py.  Simpler, because there are no tags;  we just run them all (for now).  Other stuff borrowed from tuningRunner.py and perfRunner.py in the rocMLIR repo.

File: tuna/rocmlir/load_job.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+###############################################################################
+"""
+Script for adding jobs to the MySQL database
```
Comment (pcf000) [inline tuna/rocmlir/load_job.py:28]: Adapted from tuna/miopen/subcmd/load_job.py.

File: tuna/rocmlir/load_job.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+LOGGER = setup_logger('rocmlir_load_jobs')
+
+
+def parse_args():
+  """ Argument input for the module """
+  #pylint: disable=duplicate-code
+  parser = setup_arg_parser(
+      'Insert jobs into MySQL db',
+      [TunaArgs.VERSION, TunaArgs.ARCH, TunaArgs.NUM_CU, TunaArgs.SESSION_ID])
+  parser.add_argument('-l',
```
Comment (pcf000) [inline tuna/rocmlir/load_job.py:49]: Not sure I need this any more, since I'm copying the "reason" field from the session into the job.  Or should they say different things?

File: tuna/rocmlir/rocmlir_lib.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+###############################################################################
+"""RocMLIR library, integrated with MITuna, runs 'rocminfo' cmd"""
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_lib.py]: Adapted from tuna/miopen/miopen_lib.py and tuna/example/example_lib.py.  Oops, need to update the docstring.

File: tuna/rocmlir/rocmlir_tables.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+###############################################################################
+""" The necessary tables for rocMLIR tuning integration.
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_tables.py:26]: Table definitions, already informally reviewed.  I did change a few details, like table names, that had errors.

File: tuna/rocmlir/rocmlir_tables.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+  conv_stride_w = Column(Integer, nullable=False, server_default="0")
+  dilation_h = Column(Integer, nullable=False, server_default="0")
+  dilation_w = Column(Integer, nullable=False, server_default="0")
+  group_size = Column(Integer, nullable=False, server_default="0")
+  kernel_repeats = Column(Integer, nullable=False, server_default="0")
+
+  def __repr__(self) -> str:
+    return f"ConvolutionConfig {self.to_dict()}"
+
+  options = {
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_tables.py:161]: Needs a comment.  This dict maps field names, which are also the long option names, to the short forms used in the configs.  Necessary to turn a ConvolutionConfig back into a string that we can pass to the runner.

File: tuna/rocmlir/rocmlir_tables.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    else:
+      self.logger = setup_logger('results')
+
+  @declared_attr
+  def session(self):
+    """session column"""
+    return Column(Integer, ForeignKey("session_rocmlir.id"), nullable=False)
+
+  config = Column(Integer, ForeignKey("rocmlir_conv_config.id"), nullable=False)
+  config_str = Column(Text, nullable=False)
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_tables.py:219]: The config_str field may go away again.  I added it because I get a string from the runner and needed to keep it, but then I realised that the job has the config ID and I could save that and reconstruct the string when exporting the results table.

File: tuna/rocmlir/rocmlir_tables.py [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+  perf_config = Column(Text, nullable=False)
+  kernel_tflops = Column(Float, nullable=False)
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_tables.py:222]: Used to be kernel_time.  tuningRunner.py returns TFlops of the winning perf-config and I didn't want to convert back to time, especially since it isn't a simple inverse.

File: tuna/rocmlir/rocmlir_tables.py [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+  # +++pf:  rewrite me for tuningRunner.py output!
+  def parse(self, lines):
+    """parse logger output line for result data """
+    line = lines.splitlines()[-1]
+    print(f"line being parsed is '{line}'", file=sys.stderr)
+    return line.split('\t')
+
+#pylint: disable=too-few-public-methods
+class RocMLIRDBTables(DBTablesInterface):
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_tables.py:241]: I didn't want to make another file so this class is in here.

File: tuna/rocmlir/rocmlir_worker.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+###############################################################################
+"""Builder class implements the worker interface. The purpose of this class is to run the
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_worker.py:26]: Adapted from tuna/miopen/worker/*.py and tuna/example/example_worker.py.  Need to update the docstring.

File: tuna/rocmlir/rocmlir_worker.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    obj.kernel_tflops = tflops
+
+    self.logger.info('Inserting results for job_id=%s', self.job.id)
+    query = gen_insert_query(obj, self.result_attr,
+                             self.dbt.results.__tablename__)
+    session.execute(query)
+    session.commit()
+    return True
+
+  def process_result(self, result_str: str):
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_worker.py:85]: I don't know if I *need* the process_result/update_result_table combo, but that's what miopen did and it does work.  May be a simpler way.

File: tuna/rocmlir/rocmlir_worker.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+      self.logger.info(verr)
+      self.set_job_state('errored', result=verr)
+    else:
+      if retcode != 0:
+        msg = f"Error code {retcode}, output {cmd_output}"
+        self.logger.info(msg)
+        self.set_job_state('errored', result=msg)
+      else:
+        try:
+          with open(self.output_filename(), 'r', encoding='utf8') as results:
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_worker.py:128]: tuningRunner.py writes an output file with the winning perf-config (or more than one, if given more than one conv-config, which we aren't doing here).  I haven't managed to make it output to stdout instead and thus have to open-and-read the file, and also worry about file-not-found when something weird happens and deleting the file to avoid name collisions when I reinitialise counters and flush databases.

I have hopes of cleaning this up.

File: tuna/rocmlir/rocmlir_worker.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+    env_str = " ".join(self.envmt)
+    with DbSession() as session:
+      cft = self.dbt.config_table
+      query = session.query(cft).filter(cft.id == self.job.config)
+      config = query.all()
+      if len(config) > 1:
+        raise ValueError(f"More than one config matching ID {self.job.config}")
+      config_string = config[0].config_string()
+    cmd = env_str + f" python3 ./bin/tuningRunner.py --operation conv \
+                     --config='{config_string}' --mlir-build-dir `pwd` \
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_worker.py:154]: --config is how to send tuningRunner.py just one config, instead of a file of them.
--tflops is an addition to include the winning TFlops value so I can record it.
--rocmlir_gen_flags lets me specify the GPU, which I'm only half-certain works correctly.

File: tuna/rocmlir/rocmlir_worker.py [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+                     --output={self.output_filename()} --tflops \
+                     --rocmlir_gen_flags='--device={self.gpu_id}'"
+    retcode,out = super().run_command(cmd)
+
+    return retcode,out
+
+  def get_mlir_v(self) -> str:
+    """Interface function to get mlir version info"""
+    #_, mlir_ver, _ = self.exec_docker_cmd("cat /opt/rocm/.info/version")
+    mlir_ver = "mlir_v-not-yet-implemented"
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_worker.py:165]: I really want the hash code for the revision we're working with, and I'll probably have to do "git rev-parse" in a shell call to get it.

File: tuna/rocmlir/triggers.py [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+###############################################################################
+"""Database triggers for timestamps."""
```
Comment (pcf000) [inline tuna/rocmlir/triggers.py:26]: Adapted from tuna/miopen/db/triggers.py.  I don't know if I need md5s like miopen sets up, but I did want to measure the time for each job.

**Comment by pcf000** [comment]
@cderb , can you suggest a second reviewer?  Thanks.

**Comment by pcf000** [comment]
Dammit, forgot yapf.  I'll fix that.

File: tuna/miopen/subcmd/export_db.py (hunk 2 above)
Comment (cderb) [reply tuna/miopen/subcmd/export_db.py:52]: Thanks for solving this.

File: tuna/rocmlir/load_job.py (hunk 5 above)
Comment (cderb) [reply tuna/rocmlir/load_job.py:49]: That's optional, usually I use the same reason.

File: tuna/rocmlir/rocmlir_tables.py (hunk 9 above)
Comment (cderb) [reply tuna/rocmlir/rocmlir_tables.py:219]: I've found that having the config string in the table is useful for debug interactions.

File: tuna/rocmlir/rocmlir_tables.py (hunk 10 above)
Comment (cderb) [reply tuna/rocmlir/rocmlir_tables.py:222]: Valid means of tracking performance.

File: tuna/rocmlir/rocmlir_tables.py [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+    """Set appropriate tables based on requirements"""
+    super().set_tables(sess_class)
+    # +++pf:  branch on self.config_type when we have GEMM, too.
+    self.job_table = ConvolutionJob
+    self.config_table = ConvolutionConfig
+    self.results = ConvolutionResults
+
+
+def get_tables() -> List[BASE]:
+  """Returns a list of all Example lib DB tables"""
```
Comment (cderb) [inline tuna/rocmlir/rocmlir_tables.py:269]: Old comment is here.

File: tuna/rocmlir/rocmlir_worker.py (hunk 13 above)
Comment (cderb) [reply tuna/rocmlir/rocmlir_worker.py:85]: There could be. The objective here is to package the session_retry. Which is important to survive db contention.

File: tuna/rocmlir/rocmlir_worker.py [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+  def step(self):
+    """Main functionality of the worker class. It picks up jobs in new state and executes them"""
+
+    if not self.get_job("new", "running", False):
+      #Sleep in case of DB contention
+      sleep(random.randint(1, 10))
+      return False
+
+    self.logger.info('Acquired new job: job_id=%s', self.job.id)
+    self.set_job_state('running')
```
Comment (cderb) [inline tuna/rocmlir/rocmlir_worker.py:113]: get_job will have set the state to 'running' because of the second parameter (line 109). So you don't need to run set_job_state here.

File: tuna/rocmlir/rocmlir_worker.py (hunk 15 above)
Comment (cderb) [reply tuna/rocmlir/rocmlir_worker.py:154]: The HIP_VISIBLE_DEVICES environment variable has worked in my experience.

File: tuna/rocmlir/triggers.py (hunk 17 above)
Comment (cderb) [reply tuna/rocmlir/triggers.py:26]: MIOpen uses an md5 for the config table, because there are too many columns to create the unique index. So the md5 is used for the unique key.

**Review by cderb — APPROVED** [review]
Changes to pre-existing sources are compatible with existing functionality.
New module looks to be well constructed, and suitable for an initial commit.
Will likely flag code coverage test as no unit tests have been added yet. Which an exception may need to be made for this PR.

File: tuna/miopen/subcmd/export_db.py (hunk 2 above)
Comment (pcf000) [reply tuna/miopen/subcmd/export_db.py:52]: Still not ideal, and pylint didn't like it.  Adding a docstring and a global declaration (and a suppression) satisfied pylint.  I tried a singleton object but that was worse.

**Comment by pcf000** [comment]
> coverage

Yep, there it is.  I'm looking at the existing tests and may be able to address that.

File: tuna/rocmlir/load_job.py [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+                           dbt.session_table.id == args.session_id)
+    reasons = query.all()
+    if len(reasons) > 1:
+      raise ValueError(f"More than one session matching ID {args.session_id}")
+    reason = reasons[0].reason
+
+    query = session.query(dbt.config_table.id)\
+                   .filter(dbt.config_table.valid == 1)
+    res = query.all()
+
```
Comment (msaudulhassan) [inline tuna/rocmlir/load_job.py:79]: I'd suggest renaming res to `configs`

File: tuna/rocmlir/rocmlir_tables.py [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+  tables: List[BASE] = []
+  with DbSession() as session:
+    engine = session.bind
+    connect = session.connection()
+
+    def append_if_not_exists(table):
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_tables.py:275]: There's a name clash with the Machine table, and enough of that class is in the base code that I didn't think I should change the name.  I also got tired of bumping into existing tables when testing.  Thus, append-if-not-exists to avoid double-creation.

File: tuna/rocmlir/rocmlir_tables.py [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+    append_if_not_exists(SessionRocMLIR())
+    append_if_not_exists(Machine(local_machine=True))
+    append_if_not_exists(ConvolutionConfig())
+    append_if_not_exists(ConvolutionJob())
+    append_if_not_exists(ConvolutionResults())
+
+  return tables
+
+
+def clear_tables():
```
Comment (pcf000) [inline tuna/rocmlir/rocmlir_tables.py:289]: Also for testing, ensure that the tables are in a known empty state.  Used in some tests/test*rocmlir.py files.

**Review by msaudulhassan — APPROVED** [review]
Some missing typehints

File: tuna/rocmlir/load_job.py (hunk 20 above)
Comment (pcf000) [reply tuna/rocmlir/load_job.py:79]: Renamed in working copy.
